core/trino-main/src/main/java/io/trino/operator/TopNPeerGroupLookup.java

```diff
@@ -0,0 +1,374 @@
+    }
+
+    public long size()
+    {
+        return entryCount;
+    }
+
+    public boolean isEmpty()
+    {
+        return entryCount == 0;
+    }
+
+    public long get(long groupId, long rowId)
+    {
+        checkArgument(groupId != unmappedGroupId, "Group ID cannot be the unmapped group ID");
+
+        long hash = hash(groupId, rowId);
+        long index = hash & mask;
+        if (buffer.isEmptySlot(index)) {
+            return defaultReturnValue;
+        }
+        if (hash == buffer.getPrecomputedHash(index) && equals(groupId, rowId, index)) {
+            return buffer.getValue(index);
+        }
+        // There's always an unused entry.
+        while (true) {
+            index = (index + 1) & mask;
+            if (buffer.isEmptySlot(index)) {
+                return defaultReturnValue;
+            }
+            if (hash == buffer.getPrecomputedHash(index) && equals(groupId, rowId, index)) {
+                return buffer.getValue(index);
+            }
+        }
+    }
+
+    public long put(long groupId, long rowId, long value)
+    {
+        checkArgument(groupId != unmappedGroupId, "Group ID cannot be the unmapped group ID");
+
+        long hash = hash(groupId, rowId);
+
+        long index = find(groupId, rowId, hash);
+        if (index < 0) {
+            insert(twosComplement(index), groupId, rowId, hash, value);
+            return defaultReturnValue;
+        }
+        long oldValue = buffer.getValue(index);
+        buffer.setValue(index, value);
+        return oldValue;
+    }
+
+    private long hash(long groupId, long rowId)
+    {
+        return mix(groupId * 31 + strategy.hashCode(rowId));
+    }
+
+    private boolean equals(long groupId, long rowId, long index)
+    {
+        return groupId == buffer.getGroupId(index) && strategy.equals(rowId, buffer.getRowId(index));
+    }
+
+    private void insert(long index, long groupId, long rowId, long precomputedHash, long value)
+    {
+        buffer.set(index, groupId, rowId, precomputedHash, value);
+        entryCount++;
+        if (entryCount > maxFill) {
+            rehash(bigArraySize(entryCount + 1, fillFactor));
+        }
+    }
+
+    /**
+     * Locate the index for the specified {@code groupId} and {@code rowId} key pair. If the index is unpopulated,
+     * then return the index as the two's complement value (which will be negative).
+     */
+    private long find(long groupId, long rowId, long precomputedHash)
+    {
+        long index = precomputedHash & mask;
+        if (buffer.isEmptySlot(index)) {
+            return twosComplement(index);
+        }
+        if (precomputedHash == buffer.getPrecomputedHash(index) && equals(groupId, rowId, index)) {
+            return index;
+        }
+        // There's always an unused entry.
+        while (true) {
+            index = (index + 1) & mask;
+            if (buffer.isEmptySlot(index)) {
+                return twosComplement(index);
```

Comment:
why not just return `-index`?

Reply:
Because zero is a valid index, and there is no negative equivalent of that. So you have to do something like 2s complement where everything is shifted by one.